flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -0,0 +1,203 @@
+     * @param serializationSchema serialize to transform the incoming records to {@link RawMessage}.
+     * @param metadataListener the listener for querying topic metadata.
+     * @param topicRouterProvider create related topic router to choose topic by incoming records.
+     * @param initContext context to provide information about the runtime environment.
+     */
+    public PulsarWriter(
+            SinkConfiguration sinkConfiguration,
+            PulsarSerializationSchema<IN> serializationSchema,
+            TopicMetadataListener metadataListener,
+            SerializableFunction<SinkConfiguration, TopicRouter<IN>> topicRouterProvider,
+            InitContext initContext) {
+        this.sinkConfiguration = sinkConfiguration;
+        this.deliveryGuarantee = sinkConfiguration.getDeliveryGuarantee();
+        this.serializationSchema = serializationSchema;
+        this.topicRouter = topicRouterProvider.apply(sinkConfiguration);
+        this.sinkContextAdapter = new PulsarSinkContextAdapter(initContext, sinkConfiguration);
+        this.metadataListener = metadataListener;
+        this.mailboxExecutor = initContext.getMailboxExecutor();
+
+        // Initialize topic metadata listener.
+        LOG.debug("Initialize topic metadata after creating Pulsar writer.");
+        ProcessingTimeService timeService = initContext.getProcessingTimeService();
+        metadataListener.open(sinkConfiguration, timeService);
+
+        // Initialize topic router.
+        topicRouter.open(sinkConfiguration);
+
+        // Initialize the serialization schema.
+        try {
+            InitializationContext initializationContext =
+                    initContext.asSerializationSchemaInitializationContext();
+            serializationSchema.open(initializationContext, sinkContextAdapter, sinkConfiguration);
+        } catch (Exception e) {
+            throw new FlinkRuntimeException("Cannot initialize schema.", e);
+        }
+
+        // Create this producer register after opening serialization schema!
+        this.producerRegister = new TopicProducerRegister(sinkConfiguration, serializationSchema);
+        this.pendingMessages = new Semaphore(sinkConfiguration.getMaxPendingMessages());
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        // Serialize the incoming element.
+        sinkContextAdapter.updateTimestamp(context);
+        RawMessage<byte[]> message = serializationSchema.serialize(element, sinkContextAdapter);
+
+        // Choose the right topic to send.
+        List<String> availableTopics = metadataListener.availableTopics();
+        String topic = topicRouter.route(element, message, availableTopics, sinkContextAdapter);
+
+        // Create message builder for sending message.
+        TypedMessageBuilder<?> builder = createMessageBuilder(topic, deliveryGuarantee);
```

Comment:
I am not familiar enough with Pulsar but happens with pending transactions in case the job has failed before the committables in the committer are snapshotted.

Reply:
Can you clarify this? How could the situation you mentioned happen?

Pulsar Transaction is a bit different compared to Kafka. Pulsar doesn't bind the Transaction to any Producer or Topic. You can create the transaction for your own needs. Publish or Consume messages from this transaction. The transaction is just a flag to mark the operation log.

We have to refinement the transaction usage in the Pulsar sink. Every topic would only have one live transaction between the Flink checkpoint. And since the transaction is just a mark we can reuse it by the `TxnID`.

Reply:
I think here the situation is similar to what we discussed in source connector. Basically if the messages in pending transactions are sent to Pulsar broker already (only that they are not visible to Pulsar consumers yet). If the job fails, then we lose handle to all the pending transactions. If we restart from the last checkpoint, what we can do is recreate a bunch of new transactions and resend the messages.  The only downside of this approach is the pending messages of the previous pending transactions in Pulsar broker will not be cleared until timeout.  So in this case the current implementation is safe as we are not attemping to reuse the same transactions across failures/checkpoints.

However this can raise another discussion. We previously discussed with the Pulsar transaction developers, and we find that one approach is that for each checkpoint, we will first create transactions that will be used until next checkpoint, and then store these TxnIds in the current checkpoint. In this case, lets say we have two checkpoint c1 and c2, any messages sent between c1 and c2 will use transaction id, let's say txnId1, and this txnId1 is stored in c1. Everytime we restart from last checkpoint, we can actually reuse the transactions. This way the pending messages stored in Pulsar broker will not be wasted. 

There is an edge case, so if after a restart and we get the txnId1, but this txnId1 can be timeout already (though we can set the timeout to a very large number to avoid this, but let's assume the timeout can really happen for now), in this case we can create a new transaction like the current implementation. 

So to summarize
1. current implementation is correct, the only concern is it might give pulsar broker extra burden (like the pending messages could prevent pulsar broker clear the log)
2. the proposed alternative is to alleviate the possible extra burden, and it fall back to the current implementation if the timeout still happen. The downside is it introduces state to our sink writer. 

I personally feel like. the current implementation is already good to go, the alternative can be discussed further to decide whether it is really needed